File: backend/src/test/java/com/gridsync/crdt/HybridLogicalClockTest.java
package com.gridsync.crdt;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;

class HybridLogicalClockTest {

    private UUID replicaA = UUID.randomUUID();
    private UUID replicaB = UUID.randomUUID();

    @BeforeEach
    void setUp() {
        HybridLogicalClock.reset();
    }

    @Test
    void testNowMonotonicityWhenClockJumpsBackward() {
        // Initial time
        Clock initialClock = Clock.fixed(Instant.ofEpochMilli(1000), ZoneId.of("UTC"));
        HybridLogicalClock hlc1 = HybridLogicalClock.now(replicaA, initialClock);
        assertEquals(1000L, hlc1.physicalTime());
        assertEquals(0, hlc1.logicalCounter());

        // Clock jumps backward to 900
        Clock backwardClock = Clock.fixed(Instant.ofEpochMilli(900), ZoneId.of("UTC"));
        HybridLogicalClock hlc2 = HybridLogicalClock.now(replicaA, backwardClock);
        
        // Assert monotonicity holds (physical time stays at 1000, logical counter increments)
        assertEquals(1000L, hlc2.physicalTime());
        assertEquals(1, hlc2.logicalCounter());
        assertTrue(hlc2.compareTo(hlc1) > 0);
        
        // Clock is still backward
        HybridLogicalClock hlc3 = HybridLogicalClock.now(replicaA, backwardClock);
        assertEquals(1000L, hlc3.physicalTime());
        assertEquals(2, hlc3.logicalCounter());
        assertTrue(hlc3.compareTo(hlc2) > 0);
        
        // Clock jumps forward past the previous max to 1100
        Clock forwardClock = Clock.fixed(Instant.ofEpochMilli(1100), ZoneId.of("UTC"));
        HybridLogicalClock hlc4 = HybridLogicalClock.now(replicaA, forwardClock);
        assertEquals(1100L, hlc4.physicalTime());
        assertEquals(0, hlc4.logicalCounter());
        assertTrue(hlc4.compareTo(hlc3) > 0);
    }
    
    @Test
    void testCompareTo() {
        HybridLogicalClock hlc1 = new HybridLogicalClock(1000, 0, replicaA);
        HybridLogicalClock hlc2 = new HybridLogicalClock(1000, 1, replicaA);
        HybridLogicalClock hlc3 = new HybridLogicalClock(1001, 0, replicaA);
        
        assertTrue(hlc2.compareTo(hlc1) > 0);
        assertTrue(hlc3.compareTo(hlc2) > 0);
    }
}

File: frontend/src/crdt/__tests__/HybridLogicalClock.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { HybridLogicalClock } from '../HybridLogicalClock';
import { v4 as uuidv4 } from 'uuid';

describe('HybridLogicalClock', () => {
  beforeEach(() => {
    HybridLogicalClock.reset();
  });

  it('compareTo', () => {
    const replicaA = uuidv4();
    const replicaB = uuidv4();
    const hlc1 = new HybridLogicalClock(1000, 0, replicaA);
    const hlc2 = new HybridLogicalClock(1000, 1, replicaA);
    const hlc3 = new HybridLogicalClock(1001, 0, replicaA);
    const hlc4 = new HybridLogicalClock(1000, 0, replicaB);

    expect(hlc1.compareTo(hlc1)).toBe(0);
    expect(hlc1.compareTo(hlc2)).toBe(-1);
    expect(hlc2.compareTo(hlc1)).toBe(1);
    expect(hlc1.compareTo(hlc3)).toBe(-1);
    expect(hlc3.compareTo(hlc1)).toBe(1);
    expect(hlc1.compareTo(hlc4)).toBe(replicaA > replicaB ? 1 : (replicaA < replicaB ? -1 : 0));
  });

  it('now() monotonicity when clock jumps backward', () => {
    const replicaA = uuidv4();
    let time = 1000;
    const mockClock = () => time;

    const hlc1 = HybridLogicalClock.now(replicaA, mockClock);
    expect(hlc1.physicalTime).toBe(1000);
    expect(hlc1.logicalCounter).toBe(0);

    // time jumps backward
    time = 900;
    const hlc2 = HybridLogicalClock.now(replicaA, mockClock);
    expect(hlc2.physicalTime).toBe(1000); // preserves last max
    expect(hlc2.logicalCounter).toBe(1);  // increments counter

    // time stays same
    const hlc3 = HybridLogicalClock.now(replicaA, mockClock);
    expect(hlc3.physicalTime).toBe(1000);
    expect(hlc3.logicalCounter).toBe(2);

    // time moves forward again
    time = 1005;
    const hlc4 = HybridLogicalClock.now(replicaA, mockClock);
    expect(hlc4.physicalTime).toBe(1005);
    expect(hlc4.logicalCounter).toBe(0);
  });
});
